### ISSUE TITLE
NativeAOT (net9.0) error on linux-bionic-arm/arm64

### ISSUE THREAD

**Issue by josephmoresena** [dotnet/runtime#105880]
### Description

I'm building the [NativeAOT-AndroidHelloJniLib jnetinterface branch](https://github.com/josephmoresena/NativeAOT-AndroidHelloJniLib/tree/jnetinterface) with .NET 9 preview, and I'm encountering an issue that is not present in .NET 8.

When I'm targeting linux-bionic-arm64, I get the following error from the linker:

    AndroidHelloJniLib failed with 3 error(s) (41.9s) → bin/Release/net9.0/linux-bionic-arm64/libhello-jni.dll
    ld.lld : error : relocation R_AARCH64_PREL32 cannot be used against symbol 'vtable for Rxmxnx_JNetInterface_Core_Rxmxnx_JNetInterface_Internal_INumericValue_2'; recompile with -fPIC

When I'm targeting linux-bionic-arm, I get the following runtime error:

    FATAL EXCEPTION: main
    Process: com.example.hellojni, PID: 3602
    java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "_ZTV71Rxmxnx_JNetInterface_Core_Rxmxnx_JNetInterface_Internal_INumericValue_2" referenced by "/data/app/~~NjqszTGGRScDt8U2_7MjEg==/com.example.hellojni-ZAZYn--67TH2WFmTwctWVA==/lib/arm/libhello-jni.so"...
        at java.lang.Runtime.loadLibrary0(Runtime.java:1087)
        at java.lang.Runtime.loadLibrary0(Runtime.java:1008)
        at java.lang.System.loadLibrary(System.java:1664)
        at com.example.hellojni.HelloJni.<clinit>(HelloJni.kt:73)
        at java.lang.Class.newInstance(Native Method)
        at android.app.AppComponentFactory.instantiateActivity(AppComponentFactory.java:95)
        at androidx.core.app.CoreComponentFactory.instantiateActivity(CoreComponentFactory.java:45)
        at android.app.Instrumentation.newActivity(Instrumentation.java:1253)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3353)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3601)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)

When compiling any project that references JNetInterface on osx-x64, the same error appears:

```
Rxmxnx.JNetInterface.ApplicationTest failed with 1 error(s) (4,7s) → bin/Release/net9.0/osx-x64/Rxmxnx.JNetInterface.ApplicationTest.dll
    Undefined symbols for architecture x86_64:
      "vtable for Rxmxnx_JNetInterface_Primitives_Intermediate_Rxmxnx_JNetInterface_Internal_INumericValue_2", referenced from:
          ___external_CommonFixupsTable_references in Rxmxnx.JNetInterface.ApplicationTest.o
      NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
    ld: symbol(s) not found for architecture x86_64
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

JNetInterface is a package of my own. The purpose of [the mentioned interface](https://github.com/josephmoresena/Rxmxnx.JNetInterface/blob/main/src/Intermediate/Rxmxnx.JNetInterface.Primitives.Intermediate/Internal/INumericValue.cs) is to reuse `INumberBase<TSelf>` static methods in built-in .NET numbers within JNI-specific structs in order to expose them as `INumberBase<TSelf>`.

```
namespace Rxmxnx.JNetInterface.Internal;

/// <summary>
/// This interface exposes an object that represents a java primitive number.
/// </summary>
/// <typeparam name="TPrimitive">Type of JNI primitive number.</typeparam>
/// <typeparam name="TValue">Type of the .NET equivalent structure.</typeparam>
[Browsable(false)]
[EditorBrowsable(EditorBrowsableState.Never)]
internal interface INumericValue<TPrimitive, TValue> : IBinaryNumber<TPrimitive>
	where TPrimitive : unmanaged, IPrimitiveNumericType<TPrimitive, TValue>, IPrimitiveValue<TValue>,
	IBinaryNumber<TPrimitive>, IPrimitiveEquatable
	where TValue : unmanaged, IBinaryNumber<TValue>, IMinMaxValue<TValue>, IConvertible
{
	static Int32 INumberBase<TPrimitive>.Radix => 2;

	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	static Boolean INumberBase<TPrimitive>.IsCanonical(TPrimitive value) => true;
	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	static Boolean INumberBase<TPrimitive>.IsComplexNumber(TPrimitive value) => false;
	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	static Boolean INumberBase<TPrimitive>.IsImaginaryNumber(TPrimitive value) => false;

	static Boolean INumberBase<TPrimitive>.TryConvertFromChecked<TOther>(TOther value, out TPrimitive result)
	{
		Unsafe.SkipInit(out result);
		if (typeof(TOther) != typeof(TPrimitive) && typeof(TOther) != typeof(TValue))
			return TValue.TryConvertFromChecked(value, out Unsafe.As<TPrimitive, TValue>(ref result));
		result = Unsafe.As<TOther, TPrimitive>(ref value);
		return true;
	}
	static Boolean INumberBase<TPrimitive>.TryConvertFromSaturating<TOther>(TOther value, out TPrimitive result)
	{
		Unsafe.SkipInit(out result);
		if (typeof(TOther) != typeof(TPrimitive) && typeof(TOther) != typeof(TValue))
			return TValue.TryConvertFromSaturating(value, out Unsafe.As<TPrimitive, TValue>(ref result));
		result = Unsafe.As<TOther, TPrimitive>(ref value);
		return true;
	}
	static Boolean INumberBase<TPrimitive>.TryConvertFromTruncating<TOther>(TOther value, out TPrimitive result)
	{
		Unsafe.SkipInit(out result);
		if (typeof(TOther) != typeof(TPrimitive) && typeof(TOther) != typeof(TValue))
			return TValue.TryConvertFromTruncating(value, out Unsafe.As<TPrimitive, TValue>(ref result));
		result = Unsafe.As<TOther, TPrimitive>(ref value);
		return true;
	}
	static Boolean INumberBase<TPrimitive>.TryConvertToChecked<TOther>(TPrimitive value, out TOther result)
	{
		if (typeof(TOther) != typeof(TPrimitive) && typeof(TOther) != typeof(TValue))
			return TValue.TryConvertToChecked(value.Value, out result!);
		Unsafe.SkipInit(out result);
		Unsafe.As<TOther, TValue>(ref result) = value.Value;
		return true;
	}
	static Boolean INumberBase<TPrimitive>.TryConvertToSaturating<TOther>(TPrimitive value, out TOther result)
	{
		if (typeof(TOther) != typeof(TPrimitive) && typeof(TOther) != typeof(TValue))
			return TValue.TryConvertToSaturating(value.Value, out result!);
		Unsafe.SkipInit(out result);
		Unsafe.As<TOther, TValue>(ref result) = value.Value;
		return true;
	}
	static Boolean INumberBase<TPrimitive>.TryConvertToTruncating<TOther>(TPrimitive value, out TOther result)
	{
		if (typeof(TOther) != typeof(TPrimitive) && typeof(TOther) != typeof(TValue))
			return TValue.TryConvertToTruncating(value.Value, out result!);
		Unsafe.SkipInit(out result);
		Unsafe.As<TOther, TValue>(ref result) = value.Value;
		return true;
	}
}
```

### Reproduction Steps

* Checkout jnetbranch
* Change target framework at AndroidHelloJniLib.csproj to net9.0
* Follow the steps to build libhello-jni.so

### Expected behavior

* libhello-jni.so binary builds.

### Actual behavior

* libhello-jni.o build but linker fails in linux-bionic-arm64 or builds a non-functional binary in linux-bionic-arm.

### Regression?

* Yes, with net8.0 libhello-jni.so builds well in linux-bionic-arm64.

### Known Workarounds

* Use net8.0 but only for arm64
* Do not implement the abstract members `INumberBase<TPrimitive>` in the `INumericValue<TPrimitive, TValue>` interface directly, just expose the functionality statically for inheritors of the interface to be implemented.

### Configuration

* macOS Montery 12.7.5, Intel x86-64
* .NET 9.0 Preview 6.24328.19
* Android Studio Jellyfish | 2023.3.1 Patch 1
* NDK r27

**Comment by josephmoresena** [comment]
Seems to be related to implementing interfaces abstract members on subinterfaces.

**Comment by josephmoresena** [comment]
Fixed by switching to JNetInterface version 0.4.0 where no abstract member of `INumberBase<TSelf>` is implemented in `INumericValue<TPrimitive, TValue>`.
However, this needs to be validated. This issue only occurs when compiling (using the older JNetInterface version) with NativeAOT on .NET 9.0.
